File: 0_10000/9001_10000/9252.py
def get_mat(a, b):
    mat = [[0 for i in range(len(a) + 1)] for j in range(len(b) + 1)]
    direction = [[0 for i in range(len(a) + 1)] for j in range(len(b) + 1)]

    for i in range(1, len(mat)):
        for j in range(1, len(mat[0])):
            if a[j - 1] == b[i - 1]:
                mat[i][j] = mat[i - 1][j - 1] + 1
            elif mat[i - 1][j] < mat[i][j - 1]:
                direction[i][j] = -1
                mat[i][j] = mat[i][j - 1]
            else:
                direction[i][j] = 1
                mat[i][j] = mat[i - 1][j]

    return mat, direction


a = input()
b = input()

mat, direction = get_mat(a, b)

index = mat[-1][-1]

print(index)

ans = ''

a_pos = len(a)
b_pos = len(b)

while index > 0:

    val = direction[b_pos][a_pos]

    if val == 0:
        index -= 1
        ans = a[a_pos - 1] + ans
        a_pos -= 1
        b_pos -= 1

    elif val == 1:
        b_pos -= 1
    
    else:
        a_pos -= 1


print(ans)

File: 0_10000/9001_10000/9252.cpp
#include <stdio.h>
#include <iostream>
#include <vector>
#include <string>

using namespace std;

vector<vector<int>> get_mat(std::vector<vector<int>>& mat, string a, string b){

  vector<vector<int>> dir(mat.size(), vector<int>(mat[0].size(), 0));
  for (int i = 1; i < mat.size(); i++){
    for (int j = 1; j < mat[i].size(); j++){
      if (a[i - 1] == b[j - 1]){
        mat[i][j] = mat[i - 1][j - 1] + 1;
      } 
      else if (mat[i][j - 1] < mat[i - 1][j]){
        mat[i][j] = mat[i - 1][j];
        dir[i][j] = 1;
      }
      else{
        mat[i][j] = mat[i][j - 1];
        dir[i][j] = -1;
      }
    }
  } 
  return dir;
}

string get_ans(std::vector<vector<int>> mat, string a, string b, vector<vector<int>> dir){
  string ans = "";
  int index = mat[mat.size() - 1][mat[0].size() - 1];
  int a_pos = a.size();
  int b_pos = b.size();

  while (index > 0){
    if (dir[a_pos][b_pos] == 0){
      ans = a[a_pos - 1] + ans;
      a_pos--;
      b_pos--;
      index--;
    }
    else if (dir[a_pos][b_pos] == 1){
      a_pos--;
    }
    else{
      b_pos--;
    }

  }
  return ans;
}

int main(){
  string a, b;
  cin >> a;
  cin >> b;
  vector<vector<int>> mat(a.size() + 1, vector<int>(b.size() + 1, 0));
  vector<vector<int>> dir = get_mat(mat, a, b);
  string ans = get_ans(mat, a, b, dir) ;
  int val = mat[mat.size() - 1][mat[0].size() - 1];
  cout << val << endl << ans << endl;
}

